Add Markdown support

phillipberndt opened the pull request:
This adds optional support for Markdown display. Chime supported for
some time now to prefix messages with `/md` to have them rendered as
Markdown.

The commit uses the free md4c library (https://github.com/mity/md4c/) to
render the Markdown to HTML and relies on GLib to handle rendering that.
Support for lists is currently missing (on the gtk side, the nodes aren't rendered correctly). Support is optional, if the library is missing the plugin will compile without support.

There's definitively some open questions to be addressed for this PR:

* What do I need to change to follow best practices for the autotools stuff?
* md4c is a markdown parser. Translation to HTML is taken from their examples. I hence just copied their files into a new third-party folder. Given that the code is MIT that's okay from their end, but what is this repository's policy?
* It would also be an option to copy the entire library into the tree and build a static library. What's your preference?

nacho commented:
Some points from my side here:
 - I do not like that you embed libraries here unless it is super needed.
 - I know that gnome-builder is providing already some rendering for markdown, can we use the same library they are using? this would probably be easier to integrate with that library since both programs are gtk

phillipberndt commented:
Gnome-builder uses [marked](https://github.com/chjj/marked) from what I can see. Doesn't sound like a good option given that it's written in Javascript..

I've taken a look around, and [Marker](https://github.com/fabiocolacio/Marker/) (also GTK) uses [hoedown](https://github.com/hoedown/hoedown) to translate to HTML. Looks like hoedown exports all symbols - including HTML conversion - into its shared library, so there wouldn't be a need to copy any sources into the plugin's repository.  Would that work for you?

nacho commented:
For marked I guess they use gobject introspection so yeah I guess we can skip it. So as long as it is a library that it is maintained I am fine for it. As pointed out if there is no other choice and md4c is better maintained then I would be fine with it. Though I'd rather like to see a shared library than embedding it into our source.

dwmw2 commented:
Didn't we already have an implementation of MD support?

phillipberndt commented:
I'm using the builds from your Ubuntu repository, David. If this was implemented already then the change at least didn't make it there?!

dwmw2 commented:
@jglotzer did it, although his github account no longer seems to exist. I've just pushed the latest branch I have locally, to the markdown branch at https://github.com/awslabs/pidgin-chime/commit/5afff8cd284a740950ae8d9ab65c72648bc4921c

It used libmarkdown.

phillipberndt commented:
Nice, how comes this never got merged? If there's any work pending I can look into it next week.

phillipberndt commented:
Ok, quite obvious what work was pending :) I rebased the branch onto the latest master and rewrote the parts that were overly outdated. Seems to work as expected most of the time.

Please have a look at [the latest commit](https://github.com/awslabs/pidgin-chime/pull/35/commits/291c4defd3a0f6a4710a01a09ea3825ba1915a01) though - [Pango Markup](https://developer.gnome.org/pango/stable/PangoMarkupFormat.html) and HTML seem to be too far apart to generate a good user experience. Overall it looks like the only things that are straightforward to support are

* Hyperlinks
* Bold/Italic/Underline
* So some extend code tags

Tables, lists and images aren't rendered. And whitespace isn't ignored the same way as it is in HTML files, resulting in extra space in messages.

Given that, a custom parser that tries its best to generate nice looking output for a subset of Markdown might be an alternative. (The only hard part would be footnoted hyperlinks.)

ahmetipkin commented:
Why are we not using cMark? It is official after all...

jglotzer commented:
Argggh - I changed teams at Amazon on Feb. 2 2019 and have spent the last
year pretty intensely getting up to speed in the new area (IOT).

More interestingly though in going through my Knet training I see that I
had an expired training for Github. I highly suspect that my account was
pulled due to incomplete Github Knet training. I just now completed said
training
and try to get my github account reinstated. Sorry for the confusion - I
literally just saw this today (i.e. expired training status).

John

On Fri, Feb 8, 2019 at 12:41 PM David Woodhouse <notifications@github.com>
wrote:

> @jglotzer <https://github.com/jglotzer> did it, although his github
> account no longer seems to exist. I've just pushed the latest branch I have
> locally, to the markdown branch at 5afff8c
> <https://github.com/awslabs/pidgin-chime/commit/5afff8cd284a740950ae8d9ab65c72648bc4921c>
>
> It used libmarkdown.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/awslabs/pidgin-chime/pull/35#issuecomment-461940032>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AK2LA8UewhPPf_zu0zZ8HbyaXn0EdIkpks5vLeERgaJpZM4avaFl>
> .
>

dwmw2 commented:
Some issues here. With this message:

```
/md
> Quote

Reply to quote
```
`g_markup_escape_text()` ends up converting the `>` to `%gt;` before attempting markdown, so it doesn't get handled as a quote. On the other hand if we avoid the `g_markup_escape_text()` for `/md` messages then we end up with apostrophes being converted to `&rsquo;` which Pango doesn't render, instead of `&apos;`.

Needs a little more care on the conversion, and testing.

dwmw2 commented:
I resolved the merge conflicts and pushed to a markdown branch, if anyone wants to try to fix it up.
